src/Microsoft.ML.Core/Data/Schema.cs

```diff
@@ -0,0 +1,346 @@
+
+        /// <summary>
+        /// Number of columns in the schema.
+        /// </summary>
+        public int ColumnCount => _columns.Length;
+
+        /// <summary>
+        /// Get the column by name. Throws an exception if such column does not exist.
+        /// Note that if multiple columns exist with the same name, the one with the biggest index is returned.
+        /// The other columns are considered 'hidden', and only accessible by their index.
+        /// </summary>
+        public Column this[string name]
+        {
+            get
+            {
+                Contracts.CheckValue(name, nameof(name));
+                if (!_nameMap.TryGetValue(name, out int col))
+                    throw Contracts.ExceptParam(nameof(name), $"Column '{name}' not found");
+                return _columns[col];
+            }
+        }
+
+        /// <summary>
+        /// Get the column by index.
+        /// </summary>
+        public Column this[int col]
+        {
+            get
+            {
+                Contracts.CheckParam(0 <= col && col < _columns.Length, nameof(col));
+                return _columns[col];
+            }
+        }
+
+        /// <summary>
+        /// Retrieves an index of a column by name.
+        /// </summary>
+        /// <returns>True iff the column is present in the schema.</returns>
+        public bool TryGetColumnIndex(string name, out int col) => _nameMap.TryGetValue(name, out col);
+
+        /// <summary>
+        /// Get the column by name, or <c>null</c> if the column is not present.
+        /// </summary>
+        public Column GetColumnOrNull(string name)
+        {
+            Contracts.CheckNonEmpty(name, nameof(name));
+            if (_nameMap.TryGetValue(name, out int col))
+                return _columns[col];
+            return null;
+        }
+
+        /// <summary>
+        /// This class describes one column in the schema.
+        /// </summary>
+        public sealed class Column
+        {
+            /// <summary>
+            /// The name of the column.
+            /// </summary>
+            public string Name { get; }
+
+            /// <summary>
+            /// The type of the column.
+            /// </summary>
+            public ColumnType Type { get; }
+
+            /// <summary>
+            /// The metadata of the column.
+            /// </summary>
+            public MetadataRow Metadata { get; }
+
+            public Column(string name, ColumnType type, MetadataRow metadata)
+            {
+                Contracts.CheckNonEmpty(name, nameof(name));
+                Contracts.CheckValue(type, nameof(type));
+                Contracts.CheckValueOrNull(metadata);
+
+                Name = name;
+                Type = type;
+                Metadata = metadata;
+            }
+        }
+
+        /// <summary>
+        /// The metadata of one <see cref="Column"/>.
+        /// </summary>
+        public sealed class MetadataRow
```

Comment:
>MetadataRow [](start = 28, length = 11)

You removed the interface, which is fine. But do you still want it to be `ISchematized`?

Reply:
Also the name "MetadataRow"... it is row-y in the sense that it is a collection of columns, but otherwise it doesn't seem very row-y somehow.

---
In reply to: [224834244](https://github.com/dotnet/machinelearning/pull/1167#discussion_r224834244) [](ancestors = 224834244)

Reply:
Renamed to `Metadata`

---
In reply to: [224835091](https://github.com/dotnet/machinelearning/pull/1167#discussion_r224835091) [](ancestors = 224835091,224834244)